### ISSUE TITLE
Apply EarlyStopping callback on validation set in the train function

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#2185]
Hello,

Considering the FNN scenario, when the user would like to train a forward model using a neural network, the user would like to use the `Train()` function.
For example, a user would call the function like this:
```
 double value = model.Train(dataset_.train_features(),
                               dataset_.train_labels(),
                               optimizer,
                               ens::PrintLoss(file),
                               ens::ProgressBar(),
                               ens::StoreBestCoordinates<arma::mat>());
```

Until now things are good, but if we took a closer look at the `Train` function, it is going to iterate through a set of epochs, each one with the objective is to minimize the loss. The objective of the `Train()` function is to minimize the loss as max as possible (infinity in theory). However, by doing this we are overfitting the dataset. 

To avoid overfitting we can add `ens::EarlyStopAtMinLoss(25)` callback:
```
 double value = model.Train(dataset_.train_features(),
                               dataset_.train_labels(),
                               optimizer,
                               ens::PrintLoss(file),
                               ens::ProgressBar(),
                               ens::EarlyStopAtMinLoss(25),
                               ens::StoreBestCoordinates<arma::mat>());
```
However, the training is still not perfect, since we are applying the  `ens::EarlyStopAtMinLoss(25)`  on the train features. We are not sure that the optimizer is reaching the convergence and the loss has reached the limit. A better solution might be to apply the `ens::EarlyStopAtMinLoss(25)` on the validation set rather than the training set. The main reason behind that is we need to reach the least error on what ever test set can be used.

This article [(Early stopping but when)](https://page.mi.fu-berlin.de/prechelt/Biblio/stop_tricks1997.pdf) gives the
basic way to use the Early stopping on page 2. In the actual API, the user can not
apply the Early Stopping technique on the validation set, but only on the training
set.

To improve the usage of the Early Stopping technique, we can pass the
validation set to the train function, allowing the function to use the Early Stop
only on the validation set rather than the training set. 

For example the function might look like this:
```
 double value = model.Train(dataset_.train_features(),
                               dataset_.train_labels(),
                               dataset_.validation_features(),
                               dataset_.validation_labels(),
                               optimizer,
                               ens::PrintLoss(file),
                               ens::ProgressBar(),
                               ens::EarlyStopAtMinLoss(25),
                               ens::StoreBestCoordinates<arma::mat>());
```
In this case the tain function can be used like this:
```

template<typename OutputLayerType, typename InitializationRuleType,
         typename... CustomLayers>
template<typename OptimizerType, typename... CallbackTypes>
double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Train(
      arma::mat predictors,
      arma::mat responses,
      arma::mat validatiors,
      arma::mat validators_responses,
      OptimizerType& optimizer,
      CallbackTypes&&... callbacks)
{
  ResetData(std::move(predictors), std::move(responses));

  // Train the model.
  Timer::Start("ffn_optimization");
  
  // The optimizer here is blocking until the minimum loss has been reached.
  //In this case, we have to call the evaluate from inside Optimize using *this
  
  const double out = optimizer.Optimize(*this, parameter, callbacks...);
  Timer::Stop("ffn_optimization");

  Log::Info << "FFN::FFN(): final objective of trained model is " << out
      << "." << std::endl;
  return out;
}

```
I do not know if this is possible, but I would like to call `Evaluate()` here at the end of the epoch: 

```
// Is this iteration the start of a sequence?
    if ((currentFunction % numFunctions) == 0)
    {
      terminate |= Callback::EndEpoch(*this, f, iterate, epoch++,
          overallObjective / (ElemType) numFunctions, callbacks...); 
          
       validationObjective = f.Evaluate(/* Forward the validation features and validation label */);
            
     // Print here the validation objective 
     
      // Output current objective function.
      Info << "SGD: iteration " << i << ", objective " << overallObjective
         << "." << std::endl;
    
    
      if (std::isnan(overallObjective) || std::isinf(overallObjective))
      {
        Warn << "SGD: converged to " << overallObjective << "; terminating"
            << " with failure.  Try a smaller step size?" << std::endl;

        Callback::EndOptimization(*this, f, iterate, callbacks...);
        return overallObjective;
      }
```

@zoq @rcurtin  What do you think, it is possible to realize such an implementation? is it already implemented somewhere but I have missed it?

Best regards,

**Comment by zoq** [comment]
Thanks for the detailed issue description, one solution I have in mind that doesn't involve a change to the optimizer or mlpack API is to write a custom callback function which takes the validation dataset as input. Something like this:

```c++
class EarlyStopAtMinLossOnValidation
{
 public:
  template<typename OptimizerType, typename FunctionType, typename MatType>
  bool EndEpoch(OptimizerType& /* optimizer */,
                FunctionType& function,
                const MatType& coordinates,
                const size_t /* epoch */,
                const double /* objective */)
  {
    function.Evaluate(coordinates, validationData);

    return false;
  }
};

 double value = model.Train(dataset_.train_features(),
                           dataset_.train_labels(),
                           optimizer,
                           ens::PrintLoss(file),
                           ens::ProgressBar(),
                           ens::EarlyStopAtMinLossOnValidation(dataset_.validation_features(), dataset_.validation_labels()),
                           ens::StoreBestCoordinates<arma::mat>());
```

I can see that a user might not want to use a callback in that situation, but we could implement your interface and internally use a callback to make this transparent. Let me know what you think.

**Comment by shrit** [comment]
@zoq Your solution is great.
The only thing that I miss here is the EarlyStopping because I do not see why a user would like to use it on the train data. The EarlyStopping should always be used on the validation set. Unless if it has another usage that I do know.
Both interfaces are great, we should pick the one which requires the less code to write and provide the appropriate interface.

**Comment by rcurtin** [comment]
I think a custom callback is the way to go here just like @zoq suggested. Remember that `EarlyStopping` is generic and for ensmallen, not just mlpack neural networks, so it's okay that it doesn't take an additional validation set---it wasn't made to do that.  It might be hard to find a way to provide a custom ensmallen callback that considers validation sets, but definitely not hard to just i

**Comment by rcurtin** [comment]
Sorry, wrong button...

**Comment by shrit** [comment]
Perfect, in this case, I will follow @zoq  suggestions.
I will try to create a pull request to this issue,  @rcurtin I will put the callback in the mlpack codebase.

**Comment by shrit** [comment]
@zoq @rcurtin I have been thinking about the optimizer `tolerance` in this scenario. If a user using an ANN has considered using `EarlyStoppingLossOnValidation` as a stopping criterion and forget to modify the `optimizer.Tolerance()` to `-1`. The user would not be sure which one of the criteria has triggered the end of the optimization. I think, in this case, it would be appropriate to set up the `Optimizer.Tolerance() = -1` directly from the `EarlyStoppingOnValidation` callback. What do you think??

**Comment by shrit** [comment]
Keep open

**Comment by favre49** [comment]
I had no idea that worked

**Comment by rcurtin** [comment]
mlpack-bot will actually remove the stale tag for any response (for 30 days at least).

**Comment by shrit** [comment]
Keep open

**Comment by shrit** [comment]
This issue has been addressed in https://github.com/mlpack/ensmallen/pull/165#
